votor/src/alpenglow_metrics.rs

```diff
@@ -0,0 +1,116 @@
+use {
+    histogram::Histogram,
+    solana_pubkey::Pubkey,
+    solana_votor_messages::vote::Vote,
+    std::{collections::BTreeMap, time::Duration},
+};
+
+#[derive(Debug)]
+struct VoteMetrics {
+    notar: (u64, Histogram),
+    notar_fallback: (u64, Histogram),
+    skip: (u64, Histogram),
+    skip_fallback: (u64, Histogram),
+    final_: (u64, Histogram),
+}
+
+impl VoteMetrics {
+    fn new() -> Self {
+        // TODO: pick sensible parameters for the buckets
+        let histogram = Histogram::new(10, 10).unwrap();
+        Self {
+            notar: (0, histogram.clone()),
+            notar_fallback: (0, histogram.clone()),
+            skip: (0, histogram.clone()),
+            skip_fallback: (0, histogram.clone()),
+            final_: (0, histogram),
+        }
+    }
+
+    fn record_vote(&mut self, vote: &Vote, duration: Duration) {
+        match vote {
+            Vote::Notarize(_) => {
+                self.notar.1.increment(duration.as_micros() as u64).unwrap();
+                self.notar.0 = self.notar.0.saturating_add(1);
+            }
+            Vote::NotarizeFallback(_) => {
+                self.notar_fallback
+                    .1
+                    .increment(duration.as_micros() as u64)
+                    .unwrap();
+                self.notar_fallback.0 = self.notar_fallback.0.saturating_add(1);
+            }
+            Vote::Skip(_) => {
+                self.skip.1.increment(duration.as_micros() as u64).unwrap();
+                self.skip.0 = self.skip.0.saturating_add(1);
+            }
+            Vote::SkipFallback(_) => {
+                self.skip_fallback
+                    .1
+                    .increment(duration.as_micros() as u64)
+                    .unwrap();
+                self.skip_fallback.0 = self.skip_fallback.0.saturating_add(1);
+            }
+            Vote::Finalize(_) => {
+                self.final_
+                    .1
+                    .increment(duration.as_micros() as u64)
+                    .unwrap();
+                self.final_.0 = self.final_.0.saturating_add(1);
+            }
+        }
+    }
+}
+
+/// The global metrics datastructure.
+///
+/// Each node has one of these.
+#[derive(Default)]
+pub struct AlpenglowMetrics {
+    /// Stores the node's view of how the nodes are voting.
+    vote_metrics: BTreeMap<Pubkey, VoteMetrics>,
+    /// Stores the node's view of how the nodes are producing blocks.
+    leader_metrics: BTreeMap<Pubkey, Histogram>,
+}
```

Comment:
I think if we split these out (vote/leader metrics in their own, local structs/vars), we could avoid a lot of the locking and plumbing in this PR

Reply:
Just to make sure I understand, I think you are suggesting something like the following:

```
struct VoteMetrics {
  metrics:  BTreeMap<Pubkey, VoteMetrics>,
}

struct LeaderMetrics {
  metrics: BTreeMap<Pubkey, Histogram>,
}
```

And then this way, we will just be plumbing one of the two structs through the rest of the codebase?

Reply:
yeah, that's right.

For example, I think the VoteMetrics could be created at a much lower level and avoid having to get plumbed at TPU/TVU level